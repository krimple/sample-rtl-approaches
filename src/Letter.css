.letter {
    color: blue;
    border: 2px solid blue;
    margin: 2px;
    font-size: 3em;
    background-color: orange;
    cursor: pointer;
}

.letter:disabled {
    background-color: gray;
}
