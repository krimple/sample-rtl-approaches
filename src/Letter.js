import React from 'react';
import PropTypes from 'prop-types';
import './Letter.css';

const Letter = React.memo(({addLetter, character, index, disabled}) => { // console.log(`Rendering ${index} - ${character}`);
  return (
    <button className="col-4 col-sm-2 col-md-1 btn letter"
      data-testid="letter-button"
         role="button"
      disabled={disabled}
      onClick={() => addLetter(index)}>
      {character}
    </button>
  );
});


Letter.propTypes = {
  addLetter: PropTypes.func.isRequired,
  character: PropTypes.string.isRequired,
  index: PropTypes.number.isRequired,
  disabled: PropTypes.bool
};

export default Letter;
