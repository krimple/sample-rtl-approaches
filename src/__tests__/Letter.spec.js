import { render, screen, fireEvent } from '@testing-library/react';
import Letter from '../Letter';

describe('<Letter />', () => {
  const props = {
    addLetter: jest.fn(),
    character: 'C',
    index: 3
  };

  it('renders a letter C as an enabled, unused letter', () => {
    const {container, asFragment} = render(<Letter {...props} disabled={false} />);
    expect(container).toMatchSnapshot();
    const button = screen.getByTestId('letter-button');
    expect(button).toHaveTextContent('C');
    expect(button.disabled).toBeFalsy();
  });

  it('renders a letter C as disabled, unused letter', () => {
    const {container, asFragment} = render(<Letter {...props } disabled={true} />);
    expect(container).toMatchSnapshot();
    const button = screen.getByTestId('letter-button');
    expect(button).toHaveTextContent('C');
    expect(button.disabled).toBeTruthy();
  });

  it('matches disabled snapshot', () => {
    const { asFragment } = render(<Letter {...props} disabled={true} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should call the addLetter function with the correct index', () => {
    const { container } = render(<Letter {...props}/>);
    const button = screen.getByRole('button');
    fireEvent.click(button);
    expect(props.addLetter.mock.calls).toHaveLength(1);
    expect(props.addLetter.mock.calls[0][0]).toBe(props.index);
  });

});

