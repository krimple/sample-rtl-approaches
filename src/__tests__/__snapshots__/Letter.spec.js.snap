// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Letter /> matches disabled snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn letter"
    data-testid="letter-button"
    disabled=""
  >
    C
  </button>
</DocumentFragment>
`;

exports[`<Letter /> renders a letter C as an enabled, unused letter 1`] = `
<div>
  <button
    class="btn letter"
    data-testid="letter-button"
  >
    C
  </button>
</div>
`;

exports[`<Letter /> renders a letter C as disabled, unused letter 1`] = `
<div>
  <button
    class="btn letter"
    data-testid="letter-button"
    disabled=""
  >
    C
  </button>
</div>
`;
